### PR TITLE
Issue 352 - Make URLs in Sword Module Description Clickable

### PR DESCRIPTION
**Pull request by smpurkis**
Hello,

This PR is to fix [issue 352](https://github.com/ezra-bible-app/ezra-bible-app/issues/352).

I've new to this project, so am not sure if there is functionality for clickable URLs defined somewhere else in the codebase. If there is, would someone be able to point me to that functionality.

I continued my efforts assuming this functionality wasn't in the codebase already.
I've forked and made a branch "issue-352" that adds this functionality.

The code came from https://stackoverflow.com/a/1500501 and the Regex came from https://www.codegrepper.com/code-examples/whatever/use+regex+to+get+urls+from+string.

issue-352

I'm sure there are some changes that will need to be made.

### CODE REVIEW

**Comment by smpurkis** [comment]
Accidently made "Merge branch 'master' into master", is there a quick way to fix this?

**Comment by tobias-klein** [comment]
First of all, thank you very much @smpurkis!

I just tested this with the NASB module (from Lockman foundation) and in that case there are already properly marked up URLs in the module description. I think that's a special case that we need to handle somehow.

The text looks like this before being modified:
image

And when applying the changes from this PR, this is the result:
image

Which looks like this in the popup (invalid markup + doubled URL):
image

> Accidently made "Merge branch 'master' into master", is there a quick way to fix this?

We can later squash merge the changes of this PR and then this will disappear.

File: app/frontend/components/info_popup.js [hunk 1]
```diff
@@ -121,6 +121,17 @@ class InfoPopup {
       `;
     }
 
+    function urlify(text) {
+      // replace urls in text with <a> html tag
+      // regex extracted from https://www.codegrepper.com/code-examples/whatever/use+regex+to+get+urls+from+string
+      var urlRegex = /(?:(?:https?|ftp|file):\/\/|www\.|ftp\.)(?:\([-A-Z0-9+&@#\/%=~_|$?!:,.]*\)|[-A-Z0-9+&@#\/%=~_|$?!:,.])*(?:\([-A-Z0-9+&@#\/%=~_|$?!:,.]*\)|[A-Z0-9+&@#\/%=~_|$])/igm
+      return text.replace(urlRegex, function(url) {
+        console.log(url)
+        return '<a href="' + url + '" target="_blank" title="' + url + '">' + url + '</a>';
```
Comment (tobias-klein) [inline app/frontend/components/info_popup.js:130]: 2 things

* Remove the console.log - this is only for debugging
* I would suggest to use a [template literal](https://www.w3schools.com/js/js_string_templates.asp) for putting the variables into the markup.

**Review by tobias-klein — COMMENTED** [review]
Looks fine in general. One special case is when there are already valid urls in the markup. In that case the conversion cannot be applied.

Style: Even though it's not required in JS we do ; after statements - and that's missing for line 127 and 133.

File: app/frontend/components/info_popup.js (hunk 1 above)
Comment (smpurkis) [reply app/frontend/components/info_popup.js]: - Have removed the console.log in the latest changes, ty for spotting that.
- The title attribute enables the URL to appear on hover over the link. I have removed it in latest changes. 
- Change made.

I wasn't sure the best place to put the "urlify" function, any suggestions.

**Comment by smpurkis** [comment]
I've tried to account for the
```
<a href="https://www.google.com/">https://www.google.com/</a>
```
case.

Mind retrying it now?

**Comment by tobias-klein** [comment]
Thanks for the updates @smpurkis!
It looks better now.

I have one more suggestion. Since the module description is also used in the installation assistant (see screenshot below), I would suggest to move the `urlify` function into `sword_module_helper.js` and call it directly in the context of `getModuleDescription`. Then the text will contain the links both in the installation assistant as well as in the info popup.

image

**Comment by smpurkis** [comment]
Nice, I have made the changes you suggested. I moved the `urlify` function to the bottom of `sword_module_helper.js` file, so the `module.exports...` are readable and applied `urlify` to `getModuleDescription`. 

Thank you for the warm welcome @tobias-klein! I've been looking for some open-source Christian software to contribute to for a little while. I'm glad for any opportunity to contribute and hopefully help the Church, (even in some small quality of life ways). :)

**Comment by tobias-klein** [comment]
@smpurkis Thank you once more! This is merged now. I just did some style fixes after merging and also added a block that ensures URLs are prefixed with http:// if that's missing.

**Pull request merged**